spec/models/manageiq/providers/redfish/inventory/collector/physical_infra_manager_spec.rb

```diff
@@ -0,0 +1,47 @@
+describe ManageIQ::Providers::Redfish::Inventory::Collector::PhysicalInfraManager do
+  subject(:collector) { described_class.new(nil, nil) }
+
+  context "#get_server_location" do
+    it "returns only id on missing chassis" do
+      server_data = {
+        "@odata.id" => "sid",
+        "Links"     => { "Chassis" => [] }
+      }
+      server = RedfishClient::Resource.new(nil, :content => server_data)
+      expect(collector.send(:get_server_location, server))
+        .to eq(:server_id => "sid")
+    end
+
+    it "retrieves chassis location" do
+      chassis = {
+        "Location" => { "key" => "value" },
+        "Links"    => {}
+      }
+      server_data = {
+        "@odata.id" => "sid",
+        "Links"     => { "Chassis" => [chassis] }
+      }
+      server = RedfishClient::Resource.new(nil, :content => server_data)
+      expect(collector.send(:get_server_location, server))
+        .to eq(:server_id => "sid", "key" => "value")
+    end
+
+    it "merges all parent locations" do
```

Comment:
Would it make sense to have a test with three layers, i.e. that the parent is again ContainedBy?

Reply:
Not really. We "proved" the base (server only, server + chassis) and induction (recurse on chassis) steps, so math has our back here.